File: dungeons-26.1.2/src/main/java/fishmod/features/ExplosiveShot.java
package fishmod.features;

import fishmod.features.dungeon.ChatCommandState;
import fishmod.utils.config.values.FishSettings;
import fishmod.utils.dungeon.DungeonClass;
import fishmod.utils.dungeon.Phase;
import fishmod.utils.events.Events;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.TimeUnit;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import net.minecraft.ChatFormatting;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.Gui;
import net.minecraft.network.chat.Component;

/**
 * Parses the Terminator "Explosive Shot" chat line and shows the per-enemy damage as a title.
 *
 *   "Your Explosive Shot hit 1 enemy for 472.5 damage."  → title "472.5", subtitle "Explosive Shot • 1 enemy"
 *   "Your Explosive Shot hit 8 enemies for 4,000 damage." → title "500",  subtitle "Explosive Shot • 8 enemies"
 *
 * Only active during the F7 Maxor fight ({@link Phase#inP1()}) — Terminator/Explosive Shot readings
 * aren't relevant to any other boss phase.
 *
 * The damage in the message is the TOTAL across all enemies hit; dividing by the enemy count gives
 * the per-target hit. Reads {@link Events#ON_GAME_MESSAGE} but never cancels (the chat line stays).
 */
public final class ExplosiveShot {

    private ExplosiveShot() {}

    // hit N enemy/enemies for D damage  (D may carry thousands commas and a decimal)
    private static final Pattern PATTERN = Pattern.compile(
            "Your Explosive Shot hit (\\d+) (?:enemy|enemies) for ([\\d,]+(?:\\.\\d+)?) damage");

    public static void init() {
        Events.ON_GAME_MESSAGE.register(ExplosiveShot::onMessage);
    }

    private static boolean onMessage(Component text) {
        if (text == null) return false;
        String s = text.getString();
        if (s == null) return false;

        // Phase.inP1() is the same verified Maxor-phase detection MaxorTickTimer already relies on
        // (driven by splits.json's boss chat lines) — reusing it instead of hand-matching the boss
        // taunt text here directly, since a hand-typed copy of that text previously drifted out of
        // sync with the real line and silently broke this gate (missing "!" after each "WELL").
        if (!FishSettings.explosiveShotEnabled || !Phase.inP1()) return false;
        if (s.indexOf("Explosive Shot") < 0) return false;

        Matcher m = PATTERN.matcher(s);
        if (!m.find()) return false;

        int enemies;
        double total;
        try {
            enemies = Integer.parseInt(m.group(1));
            total = Double.parseDouble(m.group(2).replace(",", ""));
        } catch (NumberFormatException e) {
            return false;
        }
        if (enemies <= 0) return false;

        double perEnemy = total / enemies;
        String dmg = formatDamage(perEnemy);
        Component title = Component.literal(dmg).withStyle(ChatFormatting.RED);
        Component subtitle = Component.literal("§7Explosive Shot §8• §f" + enemies
                + (enemies == 1 ? " enemy" : " enemies"));

        // ON_GAME_MESSAGE fires on the network thread — touch the HUD only on the client thread.
        Minecraft mc = Minecraft.getInstance();
        mc.execute(() -> {
            Gui hud = mc.gui;
            hud.setTimes(0, 25, 8); // snappy: no fade-in, ~1.25s hold, quick fade-out
            hud.setTitle(title);
            hud.setSubtitle(subtitle);
        });

        if (FishSettings.explosiveShotAnnounceParty && DungeonClass.isClass(DungeonClass.ARCHER)) {
            announceToParty(dmg, enemies);
        }
        return false; // keep the original chat line
    }

    /** Shares the same crit-hit info already shown on screen with the party. Delayed + rate-limit
     *  suppressed the same way {@code PartyCommandHandler.sendCmd} does. */
    private static void announceToParty(String dmg, int enemies) {
        String message = "Explosive Shot: " + dmg + " dmg (" + enemies + (enemies == 1 ? " enemy)" : " enemies)");
        CompletableFuture.delayedExecutor(250, TimeUnit.MILLISECONDS)
                .execute(() -> Minecraft.getInstance().execute(() -> {
                    Minecraft mc = Minecraft.getInstance();
                    if (mc.getConnection() != null) {
                        mc.getConnection().sendCommand("pc " + message);
                        ChatCommandState.lastPartyCommandAt = System.currentTimeMillis();
                    }
                }));
    }

    /** Whole numbers print with thousands separators; fractional values keep one decimal. */
    private static String formatDamage(double v) {
        if (v == Math.floor(v) && !Double.isInfinite(v)) return String.format("%,d", (long) v);
        return String.format("%,.1f", v);
    }
}

File: dungeons-26.1.2/src/main/kotlin/fishmod/features/ExplosiveShot.kt
package fishmod.features

import fishmod.utils.config.values.FishSettings
import fishmod.utils.dungeon.DungeonClass
import fishmod.utils.dungeon.Phase
import fishmod.utils.events.Events
import net.minecraft.ChatFormatting
import net.minecraft.client.Minecraft
import net.minecraft.network.chat.Component
import java.util.regex.Pattern

object ExplosiveShot {

    private val PATTERN: Pattern = Pattern.compile(
        "Your Explosive Shot hit (\\d+) (?:enemy|enemies) for ([\\d,]+(?:\\.\\d+)?) damage"
    )

    @JvmStatic
    fun init() {
        Events.ON_GAME_MESSAGE.register { text -> onMessage(text) }
    }

    private fun onMessage(text: Component?): Boolean {
        if (text == null) return false
        val s = text.string ?: return false

        if (!FishSettings.explosiveShotEnabled) return false
        if (s.indexOf("Explosive Shot") < 0) return false

        val m = PATTERN.matcher(s)
        if (!m.find()) return false

        val enemies: Int
        val total: Double
        try {
            enemies = m.group(1).toInt()
            total = m.group(2).replace(",", "").toDouble()
        } catch (e: NumberFormatException) {
            return false
        }
        if (enemies <= 0) return false

        val perEnemy = total / enemies
        val dmg = formatDamage(perEnemy)

        val mc = Minecraft.getInstance()
        // chat line shows on every Explosive Shot; title + party announce stay P1-only
        if (FishSettings.explosiveShotChatMessage) {
            val chatLine = Component.literal(
                "§7[Explosive Shot] §f$dmg §7dmg per " + (if (enemies == 1) "enemy" else "enemies") + " §8(" + enemies + ")"
            )
            mc.execute { mc.player?.sendSystemMessage(chatLine) }
        }

        if (!Phase.inP1()) return false

        if (FishSettings.explosiveShotShowTitle) {
            val title = Component.literal(dmg).withStyle(ChatFormatting.RED)
            val subtitle = Component.literal(
                "§7Explosive Shot §8• §f" + enemies + (if (enemies == 1) " enemy" else " enemies")
            )
            mc.execute {
                val hud = mc.gui
                hud.setTimes(0, 25, 8)
                hud.setTitle(title)
                hud.setSubtitle(subtitle)
            }
        }

        if (FishSettings.explosiveShotAnnounceParty && DungeonClass.isClass(DungeonClass.ARCHER)) {
            announceToParty(dmg, enemies)
        }
        return false
    }

    private fun announceToParty(dmg: String, enemies: Int) {
        val message = "Explosive Shot: $dmg dmg per enemy (" + enemies + (if (enemies == 1) " enemy)" else " enemies)")
        fishmod.utils.ChatQueue.enqueue("pc $message")
    }

    private fun formatDamage(v: Double): String {
        return when {
            v >= 1_000_000_000_000.0 -> String.format("%.1fT", v / 1_000_000_000_000.0).replace(".0", "")
            v >= 1_000_000_000.0     -> String.format("%.1fB", v / 1_000_000_000.0).replace(".0", "")
            v >= 1_000_000.0         -> String.format("%.1fM", v / 1_000_000.0).replace(".0", "")
            v >= 1_000.0             -> String.format("%.1fk", v / 1_000.0).replace(".0", "")
            v == Math.floor(v) && !v.isInfinite() -> String.format("%,d", v.toLong())
            else                     -> String.format("%,.1f", v)
        }
    }
}
